worker/src/lib/durable-objects/Wallet.ts

```diff
@@ -0,0 +1,376 @@
+      model
+    );
+
+    if (result.rowsWritten === 0) {
+      throw new Error("Unable to add to disallow list");
+    }
+  }
+
+  getDisallowList(): Set<DisallowListEntry> {
+    const result = this.ctx.storage.sql
+      .exec<{
+        provider: string;
+        model: string;
+      }>("SELECT provider, model FROM disallow_list")
+      .toArray();
+
+    return new Set(result);
+  }
+
+  isEventProcessed(eventId: string): boolean {
+    const count = this.ctx.storage.sql.exec<{count: number}>(
+      "SELECT COALESCE(count(*), 0) as count FROM processed_webhook_events WHERE id = ?", eventId
+    ).one().count;
+    return count > 0;
+  }
+
+  addCredits(amount: number, eventId: string): void {
+    const scaledAmount = amount * SCALE_FACTOR;
+    this.ctx.storage.transactionSync(() => {
+      this.ctx.storage.sql.exec(
+        "INSERT INTO credit_purchases (id, created_at, credits, reference_id) VALUES (?, ?, ?, ?)",
+        crypto.randomUUID(),
+        Date.now(),
+        scaledAmount,
+        eventId
+      );
+      this.ctx.storage.sql.exec("INSERT INTO processed_webhook_events (id, processed_at) VALUES (?, ?)", eventId, Date.now());
+    });
+  }
+
+  getCreditsPurchases(
+    page?: number,
+    pageSize?: number,
+  ): PaginatedPurchasedCredits {
+    return this.ctx.storage.transactionSync(() => {
+      const pageSizeValue = pageSize ?? 10;
+      const pageValue = page ?? 0;
+      const result = this.ctx.storage.sql.exec<{
+        id: string;
+        created_at: number;
+        credits: number;
+        reference_id: string;
+      }>(
+        "SELECT id, created_at, credits, reference_id FROM credit_purchases LIMIT ? OFFSET ?",
+        pageSizeValue,
+        pageValue * pageSizeValue
+      ).toArray();
+      const total = this.ctx.storage.sql.exec<{ total: number }>(
+        "SELECT COALESCE(COUNT(*), 0) as total FROM credit_purchases"
+      ).one().total;
+      let purchases: PurchasedCredits[] = [];
+      if (result.length > 0) {
+        purchases = result.map((row) => ({
+          id: row.id,
+          createdAt: row.created_at,
+          credits: row.credits / SCALE_FACTOR,
+          referenceId: row.reference_id,
+        }));
+      }
+      return {
+        purchases,
+        total,
+        page: pageValue,
+        pageSize: pageSizeValue,
+      } as PaginatedPurchasedCredits;
+    });
+  }
+
+  getTotalCreditsPurchased(): TotalCreditsPurchased {
+    const result = this.ctx.storage.sql.exec<{ totalCredits: number }>(
+      "SELECT COALESCE(SUM(credits), 0) as totalCredits FROM credit_purchases"
+    ).one().totalCredits;
+    return { totalCredits: result / SCALE_FACTOR };
+  }
+
+  getTotalDebits(orgId: string): { totalDebits: number, alertState: boolean } {
+    return this.ctx.storage.transactionSync(() => {
+      const debits = this.ctx.storage.sql
+        .exec<{ total: number }>("SELECT COALESCE(SUM(debits), 0) as total FROM aggregated_debits WHERE org_id = ?", orgId)
```

Comment:
this query is going to take forever at scale, I am not a huge fan of this solution